File: medium_29_uniform_data_partitioning_strategy/solution.cpp
#include <algorithm>
#include <iostream>
#include <iterator>
#include <sstream>
#include <string>
#include <vector>
#include <numeric>

using namespace std;

class Solution {
public:
    vector<int> calculate_partition_plan(const vector<int>& data_volumes) {
        if (data_volumes.empty()) {
            return {0, 0};
        }
        
        // Calculate GCD of all volumes
        int partition_size = data_volumes[0];
        for (size_t i = 1; i < data_volumes.size(); i++) {
            partition_size = gcd(partition_size, data_volumes[i]);
        }
        
        // Calculate total partitions needed
        int total_partitions = 0;
        for (int volume : data_volumes) {
            total_partitions += volume / partition_size;
        }
        
        return {partition_size, total_partitions};
    }

private:
    int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
};

void solve(istream& in = cin, ostream& out = cout) {
    int n;
    in >> n;
    
    vector<int> data_volumes(n);
    for (int i = 0; i < n; i++) {
        in >> data_volumes[i];
    }
    
    Solution solution;
    auto result = solution.calculate_partition_plan(data_volumes);
    
    out << result[0] << " " << result[1] << "\n";
}

int main() {
    solve();
    return 0;
}

File: medium_29_uniform_data_partitioning_strategy/solution.py
"""Reference implementation for the platform's official solution.

Update this file with the fully working algorithm that matches the
`description.txt` contract and is compatible with the wrappers.
"""

from __future__ import annotations

import sys
from typing import TextIO
import math


class Solution:
    def calculate_partition_plan(self, data_volumes):
        """
        :type data_volumes: List[int]
        :rtype: List[int]
        """
        if not data_volumes:
            return [0, 0]
        
        # Calculate GCD of all volumes
        partition_size = data_volumes[0]
        for volume in data_volumes[1:]:
            partition_size = math.gcd(partition_size, volume)
        
        # Calculate total partitions needed
        total_partitions = sum(volume // partition_size for volume in data_volumes)
        
        return [partition_size, total_partitions]


def solve(infile: TextIO = sys.stdin, outfile: TextIO = sys.stdout) -> None:
    """Read from ``infile`` and write the answer to ``outfile``.

    This signature enables deterministic unit testing while still working with
    Judge0 (which provides stdin/stdout). Replace the body with the final
    algorithm for the problem, including any necessary parsing.
    """

    lines = infile.read().strip().split('\n')
    if len(lines) < 2:
        return

    n = int(lines[0])
    data_volumes = list(map(int, lines[1].split()))
    
    solution = Solution()
    result = solution.calculate_partition_plan(data_volumes)
    
    outfile.write(f"{result[0]} {result[1]}\n")


if __name__ == "__main__":
    solve()